src/common/noisers.py

```diff
@@ -153,6 +153,78 @@ def __call__(self, x):
     z = self._func(x)
     return z + self._random_state.normal(size=z.shape, scale=self._sigma)
 
+class DiscreteGaussianMechanism:
+  """Transforms a function using the discrete gaussian mechanism.
+
+  If f(x) = (Z[1], Z[2], ..., Z[k]), then returns a function that computes
+  (Z'[1], Z'[2], ..., Z'[k]), where
+      Z'[i] = Z[i] + Y[i],
+      Y[i] ~ N_Z(x | sigma),
+  and N_Z(x | sigma) is given by the probability mass function defined on the
+  integers that is given by
+      N_Z(x | sigma) ~ exp(-0.5 x^2 / sigma^2) / (sigma * sqrt(2 * pi))
+
+  See:
+
+  Clément L. Canonne, Gautam Kamath, Thomas Steinke. "The Discrete Gaussian for
+  Differential Privacy" Advances in Neural Information Processing Systems 33
+  (NeurIPS 2020).
+  """
+
+  def __init__(
+    self, f, delta_f, epsilon, delta, num_queries=1, random_state=None):
+    """Instantiates a discrete gaussian mechanism.
+
+    Args:
+      f: A function which takes as input a database and which returns as output
+        a numpy array.
+      delta_f: The sensitivity paramater, e.g., the maximum value by which the
+        function can change for two databases that differ by only one row.
+      epsilon: Differential privacy parameter.
+      delta: Differential privacy parameter.
+      num_queries: The number of queries for which the mechanism is used. Note
+        that the constructed mechanism will be (epsilon, delta)-differentially
+        private when answering (no more than) num_queries queries.
+      random_state:  Optional instance of numpy.random.RandomState that is
+        used to seed the random number generator.
+    """
+    self._func = f
+    self._delta_f = delta_f
+
+    # This is only an estimate parameter using the continuous Gaussian as a
+    # a proxy. It is known that the two parameters are almost the same for
+    # a large regime of parameter; see Figure 1 in Canonne et al.'s paper.
+    # TODO: add a more rigorous parameter calculation based on privacy loss
+    # distributions.
+    self._sigma = accountant.get_smallest_gaussian_noise(
+      common.DifferentialPrivacyParameters(epsilon, delta),
+      num_queries, sensitivity=delta_f)
+    
+    self._random_state = random_state or np.random.RandomState()
+
+  def __call__(self, x):
+
+    def sample_discrete_gaussian(*unused):
+      # Use rejection sampling (Algorithm 3 in Canonne et al.) to sample a
```

Comment:
Cool algorithm. Maybe call out: use rejection sampling of discrete Laplace. Otherwise, it may seem weird to have discrete Laplace in the following.

Reply:
Added that in.